### PR TITLE
Add a Command Bar

### PR DESCRIPTION
**Pull request by jhff**
Added a Slack-like Command Bar to tackle issue #24.

Currently with a single option: open the config directory, which also tackles #66, but should be pretty easy to extend the command list from now on, e.g. change theme.

**Pressing Ctrl/Cmd + K should open it.**

halloy_hguHE9Cc0Q

I feel like placing it in the bottom looks better than at the top, unless the design is changed or we make it "pop over" the dashboard at the top. But let me know what you think.

### CODE REVIEW

**Comment by tarkah** [comment]
Code looks great! I'll defer to @casperstorm for styling, but here's some ideas:

- command bar is in floating modal with background semi transparent 
- drop down from top / center, we still probably want to semi-transparent background

In either case, I think the command bar should be Shrink.

**Comment by jhff** [comment]
@tarkah those last changes are amazing 👌

**Comment by tarkah** [comment]
> @tarkah those last changes are amazing ok_hand

Thanks!

**Comment by tarkah** [comment]
@jhff I've fixed an issue where closing the command bar would leave your buffer text input unfocused. Now, when clicking away / hitting ESC, the focused buffer has it's text input refocused.

This will be very important for chaining commands -> typing

**Comment by jhff** [comment]
Missed that, well spotted!

**Comment by casperstorm** [comment]
Looks great!
I pushed a small commit ([43f8495](https://github.com/squidowl/halloy/pull/177/commits/43f84953dc557cf6c4c0b54fe6edfb0165e95a60))

- Removed `height_margin` so we now position command bar on a offset to the first buffer on macOS. I think it gives a better "hover" effect:
- Slightly updated theme for combobox menu

Screenshot 2023-07-28 at 20 54 05

**Comment by tarkah** [comment]
@casperstorm Looks great! I say we add some initial options before merging so it's initially useful. Do we have a list of desired commands?

**Comment by casperstorm** [comment]
Here's a list of ideas I had in mind (not all has to be in this PR)

- [x] Open config directory 
- [ ] Toggle zen mode (hide side menu)
- [ ] Maximize/restore buffer
- [ ] Change buffer (populate it with our channels, eg "Change to: #halloy")
- [ ] Open new buffer (open a empty buffer)

**Comment by jhff** [comment]
Added:

- Maximize/restore buffer
- Open new buffer (open a empty buffer)
- Close buffer

@casperstorm what do you mean by:
> Change buffer (populate it with our channels, eg "Change to: #halloy")

Change currently focused buffer to a different channel?

**Comment by casperstorm** [comment]
> Change currently focused buffer to a different channel?

Yeah that's what I mean!

**Comment by jhff** [comment]
halloy_XTUQv3HEqW

File: src/screen/dashboard.rs
Comment (tarkah) [inline src/screen/dashboard.rs:312]: Nit: you can use `self.focus.take` so you don't need to set it to `None` after.

File: src/screen/dashboard/command_bar.rs
Comment (tarkah) [inline src/screen/dashboard/command_bar.rs:154]: Now that #180 is merged, there is a helper method in `src/dashboard` for getting all buffer entries, `all_buffers`. I'd rather we use that and pass in the list of buffers here so this doesn't depend on anything complex. 

Also, it includes the server & query buffer types which we probably want to include in command bar for switching

Comment (tarkah) [reply src/screen/dashboard/command_bar.rs:154]: Note, we may want to remove the `open_buffers` from `all_buffers` to generate a `closed_buffers` list to populate command bar. We shouldn't allow opening the same target in multiple buffers.

In the future, we can maybe add a `switch with <buffer>` command which handles already open buffers.

**Comment by casperstorm** [comment]
The channel switching through commandbar is great.

**Review by casperstorm — APPROVED** [review]

File: src/screen/dashboard.rs
Comment (jhff) [reply src/screen/dashboard.rs]: Done

File: src/screen/dashboard/command_bar.rs
Comment (jhff) [reply src/screen/dashboard/command_bar.rs]: Done with `all_buffers`.

Also removed buffer commands when no buffer is focused, except `New buffer`

**Review by tarkah — APPROVED** [review]
Killer feature! Thanks, @jhff

**Pull request merged**
